stm32n6: clocks updates and some tests

etienne-lms opened the pull request:
Updates on stm32n6 clock driver: add some missing macros related to CPU and device clock source + get_status handler.
Few regression tests for stm32n6 boards (nucleo_n657x0_q and stm32n6570_dk) for some core clocks and some device clocks.

etienne-lms committed:
tests: drivers: core clock controls for stm32n6 series

Add core clock configuration tests for STM32N6 series boards.
These tests covers CPU clock, system clock, PERCLK and LSE/HSE CSS
configuration.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
include: drivers: clock_control: stm32n6 CPU clock selection macros

Add STM32_CPUCLK_SRC_* macros as helper macros to select which source
clock feeds the CPU clock on a STM32N6 SoC.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
tests: drivers: device clock controls for stm32n6 series

Add device clock configuration tests for STM32N6 series boards.
These tests cover ADC, I2C and SPI, using ADC1, I2C1 and SPI5 that are
connected on nucleo_n657x0_q and stm32n6570_dk boards.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: clock_control: get_status handler for stm32n6

Add clock_control get_status handler for stm32n6 platforms.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -148,6 +148,24 @@
 
+#if defined(CONFIG_SOC_SERIES_STM32N6X)
+#if (DT_SAME_NODE(DT_CLOCKS_CTLR_BY_IDX(DT_NODELABEL(cpusw), 0), DT_NODELABEL(rcc)) && \
+     (DT_CLOCKS_CELL_BY_IDX(DT_NODELABEL(cpusw), 0, bus) == STM32_SRC_HSI))
```
I need to replace these space chars with a tab.

etienne-lms commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration_adc.c:
```diff
@@ -0,0 +1,165 @@
+#include <zephyr/logging/log.h>
+
+//#undef DT_DRV_COMPAT
+//#define DT_DRV_COMPAT st_stm32_adc
```
My bad. Leftover. I'll remove these.

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32n6_clock.h:
```diff
@@ -46,6 +46,15 @@
+#define STM32_SRC_PCLK1		(STM32_SRC_TIMG + 1)
+#define STM32_SRC_PCLK2		(STM32_SRC_PCLK1 + 1)
+#define STM32_SRC_PCLK4		(STM32_SRC_PCLK2 + 1)
+#define STM32_SRC_PCLK5		(STM32_SRC_PCLK4 + 1)
```
If you add these, you should also complete the `stm32_clock_control_get_subsys_rate` function in the clock driver to be able to compute the frequency. (I2S_CKIN would require to add a new node and a new compatible, so this could be dropped. Many series have it, but none supports it at the moment)

gautierg-st replied:
Keep it commented, indicating that it is not supported, like on some other series.

gautierg-st commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/boards/devices_on_hsi.overlay:
```diff
@@ -0,0 +1,87 @@
+};
+
+&i2c1 {
+	clocks = <&rcc STM32_CLOCK(APB1, 21)>, <&rcc STM32_SRC_HSI I2C1_SEL(5)>;
```
Technically, it should be STM32_SRC_HSI_DIV (since it is now added). This is also a good way to test HSI_DIV ;)

gautierg-st commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/boards/devices_on_hsi.overlay:
```diff
@@ -0,0 +1,87 @@
+};
+
+&adc1 {
+	clocks = <&rcc STM32_CLOCK(AHB1, 5)>, <&rcc STM32_SRC_HSI ADC12_SEL(5)>;
```
Ditto

gautierg-st commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/boards/devices_on_hsi.overlay:
```diff
@@ -0,0 +1,87 @@
+};
+
+&spi5 {
+	clocks = <&rcc STM32_CLOCK(APB2, 20)>, <&rcc STM32_SRC_HSI SPI5_SEL(5)>;
```
Ditto

gautierg-st commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/boards/devices_on_icx.overlay:
```diff
@@ -0,0 +1,111 @@
+&ic9 {
+	status = "okay";
+	pll-src = <1>;
+	ic-div = <2>;
```
In order to test more cases, you could use different PLLs for IC8, 9 and 10 (source clocks of the peripherals), as well as different dividers

etienne-lms replied:
Makes sense, i'll look at that.

etienne-lms reviewed (commented):
Thanks @gautierg-st for the comments.

gautierg-st commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -148,6 +148,24 @@
+#if defined(CONFIG_SOC_SERIES_STM32N6X)
+#if (DT_SAME_NODE(DT_CLOCKS_CTLR_BY_IDX(DT_NODELABEL(cpusw), 0), DT_NODELABEL(rcc)) && \
+	(DT_CLOCKS_CELL_BY_IDX(DT_NODELABEL(cpusw), 0, bus) == STM32_SRC_HSI))
+#define STM32_CPUCLK_SRC_HSI	1
```
HSI is the register default for the CPU clock, but I'm okay if `STM32_CPUCLK_SRC_HSI` is only set to 1 if it is explicitly set in dts. After all it could be configured differently by the BootROM, or by MCUBoot or something if it is supported one day.

gautierg-st replied:
Could still use `#elif` for subsequent checks. That would remove a few lines.

etienne-lms committed:
drivers: clock_control: add TIMG prescaler on STM32N6

Add support for TIMG clock domain as clock source on STM32N6 SoCs.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
dts: bindings: clock: stm32n6 TIMG prescaler

Add DT bindings for configuration of STM32N6 SoCs TIMG clock domain
prescaler.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: clock_control: add HSI_DIV on STM32N6

Enable support for HSI_DIV and its use as a clock source on STM32N6 SoCs.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: clock_control: add HCLKx and PCLKx source clocks on STM32N6

Enable support for HCLK, PCLK1, PCLK2, PCLK4 and PCLK5 as subsystem
clock sources identifiers on STM32N6 SoCs. HCLKx relates to the AHBx
buses clock and PCLKx relate to the APBx buses clocks.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
include: zephyr: dt-bindings: clock: stm32n6: add clock source IDs

Add some clock source identifiers that were missing for stm32n6xx
platforms.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
include: zephyr: dt-bindings: clock: stm32n6: fix refman reference

Correct references to STM32N6 series reference manual.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
include: zephyr: dt-bindings: clock: stm32u5: fix refman reference

Correct references to STM32U5 series reference manual.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
I've added some changes to consider TIMG source source for devices and add a test case (DTS overlay) for this.
`stm32_clock_control_get_subsys_rate()` is updated to handle the several alternate device source clock.
I'll updated the test use various PLLs and configurations for test using several ICx clocks as device source clocks in the test.

Note in the series that I've removed `STM32_CLOCK_BUS_APB4_2` ID in STM32N6 DT clock binding IDs that is not really applicable but preserved `STM32_CLOCK_BUS_APB1_2` that is used in many STM32 series a FDCAN source clock ID despite I think it should also be removed and replaced with `STM32_CLOCK_BUS_APB1`, at least for STM32N6 series.

gautierg-st commented on drivers/clock_control/clock_stm32_ll_n6.c:
```diff
@@ -293,7 +293,6 @@ static int stm32_clock_control_get_subsys_rate(const struct device *dev,
 		*rate = get_bus_clock(ahb_clock, STM32_APB2_PRESCALER);
 		break;
 	case STM32_CLOCK_BUS_APB4:
-	case STM32_CLOCK_BUS_APB4_2:
```
STM32_CLOCK_BUS_APB4_2 corresponds to the RCC_APB4H register (STM32_CLOCK_BUS_APB4 corresponds to RCC_APB4L). Technically speaking STM32_CLOCK_BUS_* designate register addresses. As such, they shouldn't even be used here, and we should only use STM32_SRC_*. But for peripherals that don't explicitly set a source clock, we default to their register clock to calculate their frequency (supposing that their register clock is their default source clock).

So this line should be kept for potential peripherals like that.

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32n6_clock.h:
```diff
@@ -45,26 +45,44 @@
 #define STM32_CLOCK_BUS_AHB5	0x260
+
 #define STM32_CLOCK_BUS_APB1	0x264
-#define STM32_CLOCK_BUS_APB1_2	0x268
 #define STM32_CLOCK_BUS_APB2	0x26C
-#define STM32_CLOCK_BUS_APB3	0x270
 #define STM32_CLOCK_BUS_APB4	0x274
-#define STM32_CLOCK_BUS_APB4_2	0x278
```
Do not remove. Even if not used now, they could be used in the future by new peripherals.

gautierg-st commented on drivers/clock_control/clock_stm32_ll_n6.c:
```diff
@@ -286,7 +295,6 @@ static int stm32_clock_control_get_subsys_rate(const struct device *dev,
 		*rate = ahb_clock;
 		break;
 	case STM32_CLOCK_BUS_APB1:
-	case STM32_CLOCK_BUS_APB1_2:
```
Do not remove

gautierg-st commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -163,6 +163,8 @@
 #endif
 #endif /* cpusw clk source is rcc */
+
+#define STM32_TIMG_PRESCALER	DT_PROP_OR(DT_NODELABEL(rcc), timg_prescaler, 1)
```
I think that it would be better to set the default value in the binding rather than here.

gautierg-st replied:
I mean in the dts/bindings/clock/st,stm32n6-rcc.yaml file.
It's better to add `default: 1` here where the property is described rather than in the code where it is harder to find.

etienne-lms replied:
Ok, crystal clear, thanks.

gautierg-st reviewed (changes requested):
I haven't seen a test where HSI divider is not 1. I think this could be interesting to test also.

gautierg-st reviewed (dismissed):
LGTM. Thanks!

erwango commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -148,6 +148,19 @@
 #define STM32_SYSCLK_SRC_IC2	1
 #endif
 
+#if defined(CONFIG_SOC_SERIES_STM32N6X)
```
Even if equivalent today in this case, I prefer we avoid the use of `CONFIG_SOC_SERIES_FOO` and seek for use of compatibles (`DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(rcc), st_stm32n6_rcc, okay)`).

This should ultimately help to factorize code as much as possible.

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_core/boards/clear_clocks.overlay:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
```
* Copyright (c) 2025 STMicroelectronics

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration_spi.c:
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Copyright

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration_i2c.c:
```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Copyright

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration_i2c.c:
```diff
@@ -0,0 +1,133 @@
+#include <soc.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/logging/log.h>
```
I wonder if we actually need log.h.

etienne-lms replied:
Sorry, It's a leftover from some tests I previously made. I'll remove, here and at few other places.

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration_adc.c:
```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Copyright.

erwango commented on tests/drivers/clock_control/stm32_clock_configuration/stm32n6_devices/src/test_stm32_clock_configuration.c:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Copyright.

erwango reviewed (changes requested):
Some of minor changes required otherwise LGTM

etienne-lms commented:
Comments addressed.
